icu4j/main/classes/core/src/com/ibm/icu/util/ULocale.java

```diff
@@ -1239,6 +1249,147 @@ public static String canonicalize(String localeID){
             }
         }
 
+        // If the BCP 47 primary language subtag matches the type attribute of a languageAlias
+        // element in Supplemental Data, replace the language subtag with the replacement value.
+        // If there are additional subtags in the replacement value, add them to the result, but
+        // only if there is no corresponding subtag already in the tag.
+        // Five special deprecated grandfathered codes (such as i-default) are in type attributes, and are also replaced.
+        try {
+          UResourceBundle languageAlias = UResourceBundle.getBundleInstance(ICUData.ICU_BASE_NAME,
+              "metadata", ICUResourceBundle.ICU_DATA_CLASS_LOADER)
+              .get("alias")
+              .get("language");
+          // language _ variant
+          if (!parser.getVariant().isEmpty()) {
+            String [] variants = parser.getVariant().split("_");
+            for(String variant : variants) {
+              try {
+                // Note the key in the metadata.txt is formated as langauge_variant
+                // instead of language__variant but lscvToID will generate
+                // language__variant so we have to build the string ourselves.
+                ULocale replaceLocale = new ULocale(languageAlias.get(
```

Comment:
optional/discuss: The code here looks like it works, but it also looks like it will be very slow, using the resource bundle API for alias lookups, building ULocale objects from intermediate strings only to take them apart again, allocating short-lived ULocale/StringBuilder/LocaleIDParser ... is there a good way to optimize? Maybe modify the parser rather than recreate it via ULocale?

Maybe fodder for a follow-up ticket?

Reply:
ok

Reply:
"ok"... did you submit a follow-up ticket? I don't see one in the incoming list.

Reply:
https://unicode-org.atlassian.net/browse/ICU-20970